File: frontend/src/api/agent-loop.ts
import api from './index'

export interface LoopDef {
  goal?: string
  max_iterations?: number
  max_tokens?: number
  max_cost?: number
  max_duration_sec?: number
  trigger?: { type: string; schedule?: string; event?: string }
  actions?: string[]
  [key: string]: any
}

export interface Loop {
  id: number
  workspace_id: number
  name: string
  description?: string
  loop_def: LoopDef
  version: string
  status: 'active' | 'draft' | 'archived'
  created_by_id?: number
  created_at: string
  updated_at: string
}

export interface LoopRun {
  id: number
  loop_id: number
  status: 'running' | 'completed' | 'failed' | 'escalated' | 'stopped'
  current_iteration: number
  max_iterations: number
  goal: string
  goal_metrics?: Record<string, number>
  working_memory?: Record<string, any>
  tokens_used: number
  cost_usd: number
  stopped_reason?: string
  started_at: string
  completed_at?: string
}

export interface LoopIteration {
  id: number
  loop_run_id: number
  iteration_num: number
  action_taken: { task: string; result: string }
  result_observed: Record<string, number>
  reasoning?: string
  decision: 'continue' | 'stop' | 'escalate' | 'wait'
  tokens_used: number
  duration_ms?: number
  created_at: string
}

export interface LoopRunDetail {
  run: LoopRun
  iterations: LoopIteration[]
}

export const loopApi = {
  list(workspaceId: number): Promise<Loop[]> {
    return api.get(`/workspaces/${workspaceId}/loops`).then(r => r.data)
  },

  get(workspaceId: number, loopId: number): Promise<Loop> {
    return api.get(`/workspaces/${workspaceId}/loops/${loopId}`).then(r => r.data)
  },

  create(workspaceId: number, data: { name: string; description?: string; loop_def: LoopDef }): Promise<Loop> {
    return api.post(`/workspaces/${workspaceId}/loops`, data).then(r => r.data)
  },

  update(workspaceId: number, loopId: number, data: Partial<Loop>): Promise<Loop> {
    return api.put(`/workspaces/${workspaceId}/loops/${loopId}`, data).then(r => r.data)
  },

  delete(workspaceId: number, loopId: number): Promise<void> {
    return api.delete(`/workspaces/${workspaceId}/loops/${loopId}`)
  },

  start(workspaceId: number, loopId: number): Promise<LoopRun> {
    return api.post(`/workspaces/${workspaceId}/loops/${loopId}/start`).then(r => r.data)
  },

  stop(workspaceId: number, runId: number): Promise<void> {
    return api.post(`/workspaces/${workspaceId}/loops/runs/${runId}/stop`)
  },

  getRuns(workspaceId: number, loopId: number, limit = 20): Promise<LoopRun[]> {
    return api.get(`/workspaces/${workspaceId}/loops/${loopId}/runs`, { params: { limit } }).then(r => r.data)
  },

  getRun(workspaceId: number, runId: number): Promise<LoopRunDetail> {
    return api.get(`/workspaces/${workspaceId}/loops/runs/${runId}`).then(r => r.data)
  },
}

File: backend/internal/model/agent_loop.go
package model

import (
	"encoding/json"
	"time"

	"gorm.io/gorm"
)

// Loop represents a saved Loop definition (YAML/JSON DSL).
type Loop struct {
	ID          uint64          `gorm:"primaryKey;autoIncrement" json:"id"`
	WorkspaceID uint64          `gorm:"not null;index" json:"workspace_id"`
	Name        string          `gorm:"size:255;not null" json:"name"`
	Description *string         `gorm:"type:text" json:"description,omitempty"`
	LoopDef     json.RawMessage `gorm:"type:jsonb;not null" json:"loop_def"`
	Version     string          `gorm:"size:50;default:1.0" json:"version"`
	Status      string          `gorm:"size:50;default:active" json:"status"`
	CreatedByID *uint64         `json:"created_by_id,omitempty"`
	UpdatedByID *uint64         `json:"updated_by_id,omitempty"`
	CreatedAt   time.Time       `json:"created_at"`
	UpdatedAt   time.Time       `json:"updated_at"`
	DeletedAt   gorm.DeletedAt  `gorm:"index" json:"-"`
}

func (Loop) TableName() string { return "agent_loops" }

// LoopRun represents one active/completed execution of a Loop.
type LoopRun struct {
	ID               uint64          `gorm:"primaryKey;autoIncrement" json:"id"`
	LoopID           uint64          `gorm:"not null;index" json:"loop_id"`
	Status           string          `gorm:"size:50;default:running" json:"status"`
	CurrentIteration int             `gorm:"default:0" json:"current_iteration"`
	MaxIterations    int             `gorm:"default:100" json:"max_iterations"`
	Goal             string          `gorm:"type:text;not null" json:"goal"`
	GoalMetrics      json.RawMessage `gorm:"type:jsonb" json:"goal_metrics,omitempty"`
	WorkingMemory    json.RawMessage `gorm:"type:jsonb;default:'{}'" json:"working_memory"`
	TokensUsed       int             `gorm:"default:0" json:"tokens_used"`
	CostUSD          float64         `gorm:"type:decimal(10,4);default:0" json:"cost_usd"`
	StoppedReason    *string         `gorm:"size:100" json:"stopped_reason,omitempty"`
	StartedAt        time.Time       `gorm:"autoCreateTime" json:"started_at"`
	CompletedAt      *time.Time      `json:"completed_at,omitempty"`
	UpdatedAt        time.Time       `json:"updated_at"`
}

func (LoopRun) TableName() string { return "agent_loop_runs" }

// LoopIteration records each Act->Observe->Reason cycle within a LoopRun.
type LoopIteration struct {
	ID             uint64          `gorm:"primaryKey;autoIncrement" json:"id"`
	LoopRunID      uint64          `gorm:"not null;index" json:"loop_run_id"`
	IterationNum   int             `gorm:"not null" json:"iteration_num"`
	ActionTaken    json.RawMessage `gorm:"type:jsonb;not null" json:"action_taken"`
	ResultObserved json.RawMessage `gorm:"type:jsonb;not null" json:"result_observed"`
	Reasoning      *string         `gorm:"type:text" json:"reasoning,omitempty"`
	Decision       string          `gorm:"size:50;not null" json:"decision"`
	TokensUsed     int             `gorm:"default:0" json:"tokens_used"`
	DurationMs     *int            `json:"duration_ms,omitempty"`
	CreatedAt      time.Time       `json:"created_at"`
}

func (LoopIteration) TableName() string { return "agent_loop_iterations" }
